aws/s3_test.go

```diff
@@ -0,0 +1,383 @@
+func setupNukeTests(t *testing.T) SetupInfo {
+	var s SetupInfo
+
+	region, err := getRandomRegion()
+	if err != nil {
+		assert.Fail(t, errors.WithStackTrace(err).Error())
+	}
+	s.region = region
+
+	s.awsSession, err = createSession(region)
+	if err != nil {
+		assert.Fail(t, errors.WithStackTrace(err).Error())
+	}
+
+	s.svc = s3.New(s.awsSession)
+	if err != nil {
+		assert.Fail(t, errors.WithStackTrace(err).Error())
+	}
+
+	s.bucketName = genTestBucketName()
+	return s
+}
+
+func testListS3BucketsWrapper(t *testing.T, bucketTags []map[string]string) {
+	s := setupNukeTests(t)
+
+	// Even if we nuke the bucket during our test - this will serve as a test to nuke non-existent bucket
+	// + also as our cleanup call
+
+	// Please note that we are passing the same session that was used to create the bucket
+	// This is required so that the defer cleanup call always gets the right bucket region
+	// to delete
+	defer nukeAllS3Buckets(s.awsSession, []*string{aws.String(s.bucketName)}, 1000)
+
+	awsSession, err := createSession("")
+	if err != nil {
+		assert.Failf(t, "Failed to create random session", errors.WithStackTrace(err).Error())
+	}
+
+	// Verify that - before creating bucket - it should not exist
+	//
+	// Please note that we are not reusing s.awsSession and creating a random session in a region other
+	// than the one in which the bucket is created - this is useful to test the sceanrio where the user has
+	// AWS_DEFAULT_REGION set to region x but the bucket is in region y.
+	bucketNamesPerRegion, err := getAllS3Buckets(awsSession, time.Now().Add(1*time.Hour*-1), s.bucketName)
+	if err != nil {
+		assert.Failf(t, "Failed to list S3 Buckets", errors.WithStackTrace(err).Error())
+	} else {
+		assert.NotContains(t, bucketNamesPerRegion[s.region], s.bucketName)
+	}
+
+	// Create test bucket
+	bucket := TestS3Bucket{
+		name: s.bucketName,
+	}
+	if len(bucketTags) > 0 {
+		bucket.tags = bucketTags
+	}
+	err = bucket.create(s.svc)
+	if err != nil {
+		assert.Failf(t, "Failed to create test bucket", errors.WithStackTrace(err).Error())
+	}
+
+	bucketNamesPerRegion, err = getAllS3Buckets(awsSession, time.Now().Add(1*time.Hour), s.bucketName)
+	if err != nil {
+		assert.Failf(t, "Failed to list S3 Buckets", errors.WithStackTrace(err).Error())
+	}
+
+	if len(bucketTags) > 0 {
+		if bucketTags[0]["Key"] == "cloud-nuke-excluded" && bucketTags[0]["Value"] == "true" {
+			assert.NotContains(t, bucketNamesPerRegion[s.region], aws.String(s.bucketName))
+			return
+		}
+	}
+	assert.Contains(t, bucketNamesPerRegion[s.region], aws.String(s.bucketName))
+}
+
+func TestListEmptyS3BucketWithoutFilterTag(t *testing.T) {
+	testListS3BucketsWrapper(t, []map[string]string{
+		{
+			"Key":   "testKey",
+			"Value": "testValue",
+		},
+	})
+}
+
+func TestListEmptyS3BucketWithFilterTag(t *testing.T) {
+	testListS3BucketsWrapper(t, []map[string]string{
+		{
+			"Key":   "cloud-nuke-excluded",
+			"Value": "true",
+		},
+	})
+}
```

Comment:
Can you add another test case with multiple tags, and with `cloud-nuke-excluded` not being the first element? Also, update the test function to accept a bool for determining if it should check for the bucket or not.

Reply:
Done.